File: JavaScript/devpossible-ton/src/models/TonArray.ts
/**
 * TonArray - Array model for TON
 * Copyright (c) 2024 DevPossible, LLC
 */

export class TonArray {
  private items: any[] = [];

  public add(value: any): void {
    this.items.push(value);
  }

  public get(index: number): any {
    return this.items[index];
  }

  public set(index: number, value: any): void {
    this.items[index] = value;
  }

  public length(): number {
    return this.items.length;
  }

  public toArray(): any[] {
    return [...this.items];
  }

  public toJSON(): any[] {
    return this.items.map(item => {
      if (item && typeof item.toJSON === 'function') {
        return item.toJSON();
      } else if (item && typeof item.getValue === 'function') {
        return item.getValue();
      } else {
        return item;
      }
    });
  }
}

File: src/Python/devpossible_ton/devpossible_ton/models/ton_array.py
"""
TonArray - Array model for TON
Copyright (c) 2024 DevPossible, LLC
"""

from typing import Any, List


class TonArray:
    """Array model for TON format."""

    def __init__(self):
        self.items: List[Any] = []

    def add(self, value: Any) -> None:
        """Add an item."""
        self.items.append(value)
    
    def push(self, value: Any) -> None:
        """Add an item (alias for add)."""
        self.add(value)

    def get(self, index: int) -> Any:
        """Get an item by index."""
        return self.items[index] if 0 <= index < len(self.items) else None

    def set(self, index: int, value: Any) -> None:
        """Set an item by index."""
        if 0 <= index < len(self.items):
            self.items[index] = value

    def length(self) -> int:
        """Get array length."""
        return len(self.items)
    
    def __len__(self) -> int:
        """Get array length."""
        return len(self.items)
    
    def __getitem__(self, index: int) -> Any:
        """Support subscript access like arr[index]."""
        value = self.get(index)
        # Unwrap TonValue objects for easier access
        if hasattr(value, 'get_value'):
            return value.get_value()
        return value
    
    def __setitem__(self, index: int, value: Any) -> None:
        """Support subscript assignment like arr[index] = value."""
        self.set(index, value)
    
    def __eq__(self, other) -> bool:
        """Support equality comparison with lists."""
        if isinstance(other, list):
            return self.to_json() == other
        elif isinstance(other, TonArray):
            return self.items == other.items
        return False

    def to_array(self) -> List[Any]:
        """Get as Python list."""
        return self.items.copy()

    def to_json(self) -> List[Any]:
        """Convert to JSON-serializable list."""
        result = []
        for item in self.items:
            if hasattr(item, 'to_json'):
                result.append(item.to_json())
            elif hasattr(item, 'get_value'):
                result.append(item.get_value())
            else:
                result.append(item)
        return result
